### PR TITLE
Add integration tests for the gardenlet landscaper

### PR DESCRIPTION
**Pull request by danielfoehrKn**
**How to categorize this PR?**

/area landscaper
/kind enhancement

**What this PR does / why we need it**:
Add integration tests for the gardenlet landscaper.
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

You can test the integration test by running the test against a cluster that has a Landscaper installed.
You can use the Gardenlet landscaper image & component descriptor that I already build.
(EDIT: after addressing the review comments, the resource name of blueprint in the component descriptor needs to be called "gardenlet-blueprint" now, so the current component descriptor does not work any more)

Use those flags when executing the test in `test/integration/landscaper/gardenlet`
```
-landscaper-kubecfg-path=<path-to-landscaper-kubecfg>
-kubecfg=<path-to-garden-cluster-kubeconfig>
-seed-kubecfg-path=<path-to-seed-cluster-kubeconfig>
-repository-context-url=eu.gcr.io/gardener-project/landscaper/dev/components
-component-name=github.com/gardener/gardener
-component-version=v1.19.0-d060239-integration
-landscaper-target-namespace=ls-system
```

**Release note**:

```other operator

```

### CODE REVIEW

File: test/framework/landscaper/commonframework.go [hunk 1]
```diff
@@ -0,0 +1,365 @@
+package landscaper
```
Comment (rfranzke) [inline test/framework/landscaper/commonframework.go:15]: license header

File: test/framework/landscaper/gardenletframework.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
+package landscaper
```
Comment (rfranzke) [inline test/framework/landscaper/gardenletframework.go:15]: license header

File: landscaper/utils/helper/helper.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helper
+
+import landscaperv1alpha1 "github.com/gardener/landscaper/apis/core/v1alpha1"
+
+// GetCondition returns the condition with the given <conditionType> out of the list of <conditions>.
+// In case the required type could not be found, it returns nil.
+func GetCondition(conditions []landscaperv1alpha1.Condition, conditionType landscaperv1alpha1.ConditionType) *landscaperv1alpha1.Condition {
```
Comment (schrodit) [inline landscaper/utils/helper/helper.go]: the landscaper helper function could be used https://github.com/gardener/landscaper/blob/master/apis/core/v1alpha1/helper/helpers.go#L47

File: test/framework/landscaper/commonframework.go [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+type LandscaperCommonConfig struct {
+	KubeconfigPathLandscaperCluster string
+	TargetNamespace                 string
+	ComponentConfig                 ComponentConfig
+}
+
+// NewLandscaperCommonFramework creates a new landscaper common framework.
+func NewLandscaperCommonFramework(cfg *LandscaperCommonConfig) *CommonFrameworkLandscaper {
+	suffix, err := utils.GenerateRandomStringFromCharset(5, "0123456789abcdefghijklmnopqrstuvwxyz")
+	gomega.Expect(err).ToNot(gomega.HaveOccurred())
```
Comment (schrodit) [inline test/framework/landscaper/commonframework.go]: nit
```suggestion
         framework.ExpectNoError(err)
```

**Comment by danielfoehrKn** [comment]
@schrodit thanks, I updated it. PTAL

**Comment by danielfoehrKn** [comment]
What is the command to retrigger the CI again? I think it is a transient error.

File: .test-defs/GardenletLandscaper.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  args:
+  - >-
+    go test -timeout=0 -mod=vendor ./test/integration/landscaper/gardenlet
+    --v -ginkgo.v -ginkgo.progress -ginkgo.noColor
+    -verbose=debug
+    -kubecfg=$TM_KUBECONFIG_PATH/gardener.config
+    -seed-kubecfg-path=$TM_KUBECONFIG_PATH/seed.config
+    -landscaper-kubecfg-path=$TM_KUBECONFIG_PATH/landscaper.config
+    -landscaper-target-namespace=$LANDSCAPER_NAMESPACE
+    -componentBaseUrl=$COMPONENT_DESCRIPTOR_BASE_URL
```
Comment (schrodit) [inline .test-defs/GardenletLandscaper.yaml]: lets keep the naming in line with the CICD naming
```suggestion
    -repositoryContextUrl=$REPOSITORY_CTX_BASE_URL
```

File: test/framework/landscaper/commonframework.go [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+	f.LandscaperClient = client.Client()
+}
+
+// RegisterLandscaperCommonFrameworkFlags adds all flags that are needed to configure a landscaper common framework to the provided flagset.
+func RegisterLandscaperCommonFrameworkFlags() *LandscaperCommonConfig {
+	newCfg := &LandscaperCommonConfig{}
+
+	flag.StringVar(&newCfg.KubeconfigPathLandscaperCluster, "landscaper-kubecfg-path", "", "the path to the kubeconfig  of the landscaper cluster that will be used for integration tests")
+	flag.StringVar(&newCfg.TargetNamespace, "landscaper-target-namespace", "ls-system", "the namespace to create the landscaper resources in")
+	flag.StringVar(&newCfg.ComponentConfig.baseUrl, "componentBaseUrl", "", "the base url of the component descriptor (e.g. eu.gcr.io/gardener-project/landscaper/dev/components)")
```
Comment (schrodit) [inline test/framework/landscaper/commonframework.go]: see comment above

File: test/framework/landscaper/commonframework.go [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+			Name:      name,
+			Namespace: f.Config.TargetNamespace,
+		},
+		Spec: landscaperv1alpha1.InstallationSpec{
+			Blueprint: landscaperv1alpha1.BlueprintDefinition{
+				Reference: &landscaperv1alpha1.RemoteBlueprintReference{
+					// the component descriptor has to contain the blueprint as a resource
+					// with type:blueprint and name: blueprint
+					// in the future, this could be made configurable
+					ResourceName: "blueprint",
```
Comment (schrodit) [inline test/framework/landscaper/commonframework.go]: think we should name the blueprint `gardenlet-blueprint` as there could be multiple blueprints in teh future

File: test/framework/landscaper/commonframework.go [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+		err = f.LandscaperClient.Delete(ctx, target)
+		if err != nil {
+			f.Logger.Debugf("Could not delete target (%s/%s): %s", target.Namespace, target.Name, err.Error())
+			return retry.MinorError(err)
+		}
+		return retry.Ok()
+	})
+}
+
+// CreateInstallation creates a new config map with retries
```
Comment (schrodit) [inline test/framework/landscaper/commonframework.go]:
```suggestion
// CreateConfigMap creates a new config map with retries
```

File: test/framework/landscaper/gardenletframework.go [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+	}
+
+	if err := f.createInstallationAndWaitToBecomeHealthy(ctx, installation); err != nil {
+		return nil, err
+	}
+
+	return installation, nil
+}
+
+// DeleteInstallationResources deletes the previously created kandscaper resources
```
Comment (schrodit) [inline test/framework/landscaper/gardenletframework.go]:
```suggestion
// DeleteInstallationResources deletes the previously created landscaper resources
```

**Review by schrodit — CHANGES_REQUESTED** [review]

File: .test-defs/GardenletLandscaper.yaml (hunk 5 above)
Comment (danielfoehrKn) [reply .test-defs/GardenletLandscaper.yaml]: thanks, additionally I renamed all the flags to remove the camel casing.

File: test/framework/landscaper/commonframework.go (hunk 7 above)
Comment (danielfoehrKn) [reply test/framework/landscaper/commonframework.go]: makes sense.

**Comment by danielfoehrKn** [comment]
@schrodit , thanks I addressed the review comments PTAL

**Review by schrodit — DISMISSED** [review]
lgtm

**Comment by danielfoehrKn** [comment]
@rfranzke if you are also fine with the PR then I would like to get it in

**Review by rfranzke — APPROVED** [review]

**Pull request merged**